### PR TITLE
Delay loading of embedded file data until static constructors are run

### PR DESCRIPTION
**Pull request by sbc100**
This code was relying on static data relocations being applied during
module instantiation, but a recent change to llvm reverts to the
previous behaviour of applying relocation during `__wasm_call_ctors`:
https://reviews.llvm.org/D127333

This change delays the loading of embedded files until static ctor time
after the relocations have been applied.

As a followup I think we can look into completely removing the pre-js
in this case.

### CODE REVIEW

**Review by dschuff — COMMENTED** [review]
this... looks fine? although I don't really have a good background knowledge of the file packager and file embedding.

File: tools/file_packager.py [hunk 1]
```diff
@@ -655,35 +669,10 @@ def generate_js(data_target, data_files, metadata):
         new DataRequest(files[i]['start'], files[i]['end'], files[i]['audio'] || 0).open('GET', files[i]['filename']);
       }\n''' % (create_preloaded if options.use_preload_plugins else create_data)
 
-  if options.has_embedded:
-    if options.obj_output:
-      if options.wasm64:
-        code += '''\
```
Comment (dschuff) [inline tools/file_packager.py:661]: so does this code go directly into the wrapper JS code? and this PR is moving it to regular library code?

File: tools/file_packager.py [hunk 2]
```diff
@@ -290,18 +290,32 @@ def generate_object_file(data_files):
...
+      .section .text,"",@
+      init_file_data:
+        .functype init_file_data () -> ()
+        global.get __emscripten_embedded_file_data@GOT
+        call _emscripten_fs_load_embedded_files
+        end_function
+
+      .section  .init_array.100,"",@
+      .p2align  2
+      .int32 init_file_data
```
Comment (kripken) [inline tools/file_packager.py:308]: How do we ensure that this global ctor runs before user ones? Is it the `100` constant in there perhaps?

Comment (sbc100) [reply tools/file_packager.py:310]: Ah yes, sorry I forgot to allocate a specific ordering.   Fixed.

File: tools/file_packager.py (hunk 1 above)
Comment (sbc100) [reply tools/file_packager.py:661]: Yes, it moved from pre-js code that would run before `__wasm_call_ctors` into a JS library function that gets called from a ctor at priority 49.

**Comment by dschuff** [comment]
In "regular" systems, I guess it must be possible to open and read files from C++ constructors, right? So one consequence of this is that this becomes no longer possible for embedded files, assuming it was before? (That's not necessarily a problem, but we should maybe put something in the doc or the changelog?)

**Comment by sbc100** [comment]
> In "regular" systems, I guess it must be possible to open and read files from C++ constructors, right? So one consequence of this is that this becomes no longer possible for embedded files, assuming it was before? (That's not necessarily a problem, but we should maybe put something in the doc or the changelog?)

No, that still works since I chose a system level priority, it will run before any user-defined constructors.   See `system/lib/README.md`

**Review by dschuff — APPROVED** [review]
got it, thanks for the explanation.

**Pull request merged**